core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,252 @@
+    staking_path = catalyst_registration_parameters.staking_path
+    reward_address = derive_human_readable_address(
+        keychain,
+        catalyst_registration_parameters.reward_address_parameters,
+        protocol_magic,
+        network_id,
+    )
+    nonce = catalyst_registration_parameters.nonce
+
+    await confirm_catalyst_registration(
+        ctx, encoded_public_key, staking_path, reward_address, nonce
+    )
+
+
+def get_auxiliary_data_cbor(
+    keychain: seed.Keychain,
+    auxiliary_data: CardanoTxAuxiliaryDataType,
+    protocol_magic: int,
+    network_id: int,
+) -> bytes:
+    if auxiliary_data.type == CardanoAuxiliaryDataType.BLOB:
+        # ensured by _validate_auxiliary_data
+        assert auxiliary_data.blob is not None
+
+        return auxiliary_data.blob
+    elif auxiliary_data.type == CardanoAuxiliaryDataType.TUPLE:
+        # ensured by _validate_auxiliary_data
+        assert auxiliary_data.metadata is not None
+        assert auxiliary_data.metadata.catalyst_registration_parameters is not None
+
+        cborized_catalyst_registration = _cborize_catalyst_registration(
+            keychain,
+            auxiliary_data.metadata.catalyst_registration_parameters,
+            protocol_magic,
+            network_id,
+        )
+
+        return cbor.encode(_wrap_metadata(cborized_catalyst_registration))
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _cborize_catalyst_registration(
+    keychain: seed.Keychain,
+    catalyst_registration_parameters: CardanoCatalystRegistrationParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> CatalystRegistration:
+    staking_node = keychain.derive(catalyst_registration_parameters.staking_path)
+    staking_key = remove_ed25519_prefix(staking_node.public_key())
+
+    catalyst_registration_payload: CatalystRegistrationPayload = {
+        1: catalyst_registration_parameters.voting_public_key,
+        2: staking_key,
+        3: derive_address_bytes(
+            keychain,
+            catalyst_registration_parameters.reward_address_parameters,
+            protocol_magic,
+            network_id,
+        ),
+        4: catalyst_registration_parameters.nonce,
+    }
+
+    catalyst_registration_payload_signature = _sign_catalyst_registration_payload(
+        keychain,
+        catalyst_registration_payload,
+        catalyst_registration_parameters.staking_path,
+    )
+    catalyst_registration_signature = {1: catalyst_registration_payload_signature}
+
+    return {
+        CATALYST_REGISTRATION_KEY: catalyst_registration_payload,
+        CATALYST_REGISTRATION_SIGNATURE_KEY: catalyst_registration_signature,
+    }
+
+
+def _sign_catalyst_registration_payload(
+    keychain: seed.Keychain,
+    catalyst_registration_payload: CatalystRegistrationPayload,
+    path: list[int],
+) -> bytes:
+    node = keychain.derive(path)
+
+    encoded_catalyst_registration = cbor.encode(
+        {CATALYST_REGISTRATION_KEY: catalyst_registration_payload}
```

Comment:
I've restructured it a bit, hopefully making it a bit clearer. https://github.com/vacuumlabs/trezor-firmware/pull/57/commits/de6ce66e59dc846769a71ec5ce9fa3e0583d45a1#diff-449dbb3cc07455194600fd241faad462fb0b7417008b0b99fef57c4c7966745bR254

Reply:
After an offline discussion with @refi93 we decided to go back to building the dict before signing, but the function has been renamed to `_create_catalyst_registration_payload_signature` so that it's clear that it's doing more than just simple signing of data.